File: Pod/Classes/Stack.m
/*
   Copyright (c) 2015 Shaps Mohsenin. All rights reserved.
 
 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions are met:
 
 1. Redistributions of source code must retain the above copyright notice, this
 list of conditions and the following disclaimer.
 
 2. Redistributions in binary form must reproduce the above copyright notice,
 this list of conditions and the following disclaimer in the documentation
 and/or other materials provided with the distribution.
 
 THIS SOFTWARE IS PROVIDED BY Shaps Mohsenin `AS IS' AND ANY EXPRESS OR
 IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
 EVENT SHALL Shaps Mohsenin OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#import "Stack.h"
#import "SPXDefines.h"

static NSMutableDictionary * __registeredStacks;
NSString *const __stackThreadContextKey = @"__stackContextKey";
NSString *const __stackTransactionKey = @"__stackTransactionKey";


@interface StackTransaction (Private)
@property (nonatomic, weak) Stack *stack;
@property (nonatomic, strong) NSManagedObjectContext *managedObjectContext;
@property (nonatomic, copy) void (^transactionBlock)();
@property (nonatomic, copy) void (^transactionCompletionBlock)();
@property (nonatomic, copy) void (^saveSynchronously)();
@property (nonatomic, copy) void (^saveAsynchronously)();
@end


@interface StackQuery (Private)
@property (nonatomic, weak) Stack *stack;
@property (nonatomic, strong) NSManagedObjectContext *managedObjectContext;
+ (instancetype)queryForManagedObjectClass:(Class)managedObjectClass entityName:(NSString *)entityName;
@end


@interface Stack ()

@property (nonatomic, strong) NSDictionary *entityNameToClassMappings;
@property (nonatomic, strong) NSPersistentStoreCoordinator *coordinator;
@property (nonatomic, strong) NSManagedObjectModel *managedObjectModel;

@property (nonatomic, strong) NSManagedObjectContext *rootContext;
@property (nonatomic, strong) NSManagedObjectContext *mainThreadContext;
@property (nonatomic, strong) NSManagedObjectContext *transactionContext;
@property (nonatomic, strong) NSManagedObjectContext *currentThreadContext;

@end

@implementation Stack

@synthesize currentThreadContext = _currentThreadContext;

#pragma mark - Initializers

- (void)loadEntityMappings
{
  NSManagedObjectModel *model = self.managedObjectModel;
  NSMutableDictionary *mappings = [NSMutableDictionary new];
  
  [model.entitiesByName enumerateKeysAndObjectsUsingBlock:^(NSString *entityName, NSEntityDescription *entityDescription, BOOL *stop) {
    mappings[entityDescription.managedObjectClassName] = entityName;
  }];

  self.entityNameToClassMappings = mappings.copy;
}

+ (instancetype)defaultStack
{
  static Stack *_defaultStack = nil;
  static dispatch_once_t oncePredicate;
  dispatch_once(&oncePredicate, ^{
    NSString *name = [NSBundle mainBundle].infoDictionary[@"CFBundleName"];
    name = [name stringByAppendingString:@"_Default"];
    _defaultStack = [Stack registerStackWithName:name model:Stack.defaultStackModel storeURL:Stack.defaultStackURL inMemoryOnly:NO];
  });
  
  return _defaultStack;
}

+ (instancetype)memoryStack
{
  static Stack *_memoryStack = nil;
  static dispatch_once_t oncePredicate;
  dispatch_once(&oncePredicate, ^{
    NSString *name = [NSBundle mainBundle].infoDictionary[@"CFBundleName"];
    name = [name stringByAppendingString:@"_Memory"];
    _memoryStack = [Stack registerStackWithName:name model:Stack.defaultStackModel storeURL:Stack.defaultStackURL inMemoryOnly:YES];
  });
  
  return _memoryStack;
}

+ (instancetype)stackNamed:(NSString *)name
{
  return __registeredStacks[name];
}

+ (instancetype)registerStackWithName:(NSString *)name model:(NSManagedObjectModel *)model storeURL:(NSURL *)storeURL inMemoryOnly:(BOOL)memoryOnly
{
  SPXAssertTrueOrReturnNil(!__registeredStacks[name]);
  
  if (!__registeredStacks) {
    __registeredStacks = [NSMutableDictionary new];
  }
  
  if (__registeredStacks[name]) {
    return __registeredStacks[name];
  }
  
  Stack *stack = [[Stack alloc] initWithName:name model:model storeURL:storeURL inMemoryOnly:memoryOnly];
  __registeredStacks[name] = stack;
  
  return stack;
}

- (instancetype)initWithName:(NSString *)name model:(NSManagedObjectModel *)model storeURL:(NSURL *)storeURL inMemoryOnly:(BOOL)memoryOnly
{
  self = [super init];
  SPXAssertTrueOrReturnNil(self);
  
  _managedObjectModel = model;
  _coordinator = [[NSPersistentStoreCoordinator alloc] initWithManagedObjectModel:model];
  SPXAssertTrueOrReturnNil(_coordinator);
  
  NSError *error = nil;
  NSDictionary *autoMigratingOptions = self.class.autoMigratingOptions;
  NSPersistentStore *store = nil;
  
  if (memoryOnly) {
    store = [_coordinator addPersistentStoreWithType:NSInMemoryStoreType configuration:nil URL:nil options:autoMigratingOptions error:&error];
  } else {
    store = [_coordinator addPersistentStoreWithType:NSSQLiteStoreType configuration:nil URL:storeURL options:autoMigratingOptions error:&error];
  }
  
  if (error) {
    SPXLog(@"%@", error);
    error = nil;
  }
  
  if (!store && !memoryOnly) {
    if (![[NSFileManager defaultManager] removeItemAtURL:storeURL error:&error]) {
      SPXLog(@"%@", error);
    }
    
    store = [_coordinator addPersistentStoreWithType:NSSQLiteStoreType configuration:nil URL:storeURL options:autoMigratingOptions error:&error];
    
    if (error) {
      SPXLog(@"%@", error);
    }
    
    SPXAssertTrueOrReturnNil(store);
  }
  
  _rootContext = [[NSManagedObjectContext alloc] initWithConcurrencyType:NSPrivateQueueConcurrencyType];
  _rootContext.persistentStoreCoordinator = _coordinator;
  [self loadEntityMappings];
  
  return self;
}

#pragma mark - Helpers

+ (NSManagedObjectModel *)defaultStackModel
{
  return [NSManagedObjectModel mergedModelFromBundles:nil];
}

+ (NSURL *)defaultStackURL
{
  NSURL *url = [[[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask] lastObject];
  NSString *storeName = [NSBundle mainBundle].infoDictionary[@"CFBundleName"];
  return [url URLByAppendingPathComponent:[NSString stringWithFormat:@"%@.sqlite", storeName]];
}

+ (NSDictionary *)autoMigratingOptions
{
  return @{  NSMigratePersistentStoresAutomaticallyOption : @(YES),
             NSInferMappingModelAutomaticallyOption       : @(YES),
             };
}

+ (void)handleError:(NSError *)error
{
  if (!error) {
    return;
  }
  
  for (NSArray *detailedError in error.userInfo.allValues) {
    if ([detailedError isKindOfClass:[NSArray class]]) {
      for (NSError *err in detailedError) {
        if ([err respondsToSelector:@selector(userInfo)]) {
          SPXLog(@"Error Details: %@", err.userInfo);
        } else {
          SPXLog(@"Error Details: %@", err);
        }
      }
    } else {
      SPXLog(@"Error: %@", detailedError);
    }
  }
  
  SPXLog(@"Error Message: %@", [error localizedDescription]);
  SPXLog(@"Error Domain: %@", [error domain]);
  SPXLog(@"Recovery Suggestion: %@", [error localizedRecoverySuggestion]);
}

#pragma mark - Contexts

- (NSManagedObjectContext *)mainThreadContext
{
  return _mainThreadContext ?: ({
    NSManagedObjectContext *mainThreadContext = [[NSManagedObjectContext alloc] initWithConcurrencyType:NSMainQueueConcurrencyType];
    mainThreadContext.parentContext = self.rootContext;
    _mainThreadContext = mainThreadContext;
  });
}

- (void)setCurrentThreadContext:(NSManagedObjectContext *)context
{
  NSThread *thread = NSThread.currentThread;
  
  if (context) {
    thread.threadDictionary[__stackThreadContextKey] = context;
  } else {
    [thread.threadDictionary removeObjectForKey:__stackThreadContextKey];
  }
}

- (NSManagedObjectContext *)currentThreadContext
{
  NSManagedObjectContext *context = nil;
  NSThread *thread = NSThread.currentThread;
  
  context = thread.threadDictionary[__stackThreadContextKey];
  
  if (!context) {
    if (NSThread.isMainThread) {
      context = self.mainThreadContext;
    } else {
      context = [[NSManagedObjectContext alloc] initWithConcurrencyType:NSPrivateQueueConcurrencyType];
      context.parentContext = self.mainThreadContext;
    }
    
    self.currentThreadContext = context;
  }
  
  return context;
}

- (NSManagedObjectContext *)transactionContext
{
  NSManagedObjectContext *context = [[self transactionStack].lastObject managedObjectContext];
  
  if (!context) {
    context = [[NSManagedObjectContext alloc] initWithConcurrencyType:NSPrivateQueueConcurrencyType];
    context.parentContext = self.mainThreadContext;
    [self.transactionStack.lastObject setManagedObjectContext:context];
  }

  return context;
}

#pragma mark - Deletes

- (void (^)(NSArray *))deleteObjects
{
  return ^(NSArray *objects) {
    for (id object in objects) {
      SPXAssertTrueOrReturn([object isKindOfClass:[NSManagedObject class]]);
      [self.currentThreadContext deleteObject:object];
    }
  };
}

- (void (^)(NSManagedObjectID *))deleteWhereObjectID
{
  return ^(NSManagedObjectID *objectID) {
    [self.currentThreadContext deleteObject:[self.transactionContext objectWithID:objectID]];
  };
}

#pragma mark - Queries

- (NSString *(^)(__unsafe_unretained Class))entityNameForClass
{
  return ^(__unsafe_unretained Class managedObjectClass) {
    return self.entityNameToClassMappings[NSStringFromClass(managedObjectClass)];
  };
}

- (StackQuery *(^)(__unsafe_unretained Class))query
{
  return ^(__unsafe_unretained Class managedObjectClass) {
    SPXAssertTrueOrPerformAction([managedObjectClass isSubclassOfClass:[NSManagedObject class]], return (StackQuery *)nil);
    StackQuery *query = [StackQuery queryForManagedObjectClass:managedObjectClass entityName:self.entityNameForClass(managedObjectClass)];
    query.stack = self;
    return query;
  };
}

#pragma mark - Transactions

- (NSMutableArray *)transactionStack
{
  NSThread *thread = NSThread.currentThread;
  NSMutableArray *stack = thread.threadDictionary[__stackTransactionKey];
  
  if (!stack) {
    stack = [NSMutableArray new];
    thread.threadDictionary[__stackTransactionKey] = stack;
  }
  
  return stack;
}

- (StackTransaction *)pushTransaction
{
  NSMutableArray *stack = [self transactionStack];
  StackTransaction *transaction = [StackTransaction new];
  transaction.stack = self;
  [stack addObject:transaction];
  return transaction;
}

- (void)popTransaction
{
  NSMutableArray *stack = [self transactionStack];
  [stack removeLastObject];
}

- (void (^)(void (^transactionBlock)()))transaction
{  
  __weak typeof(self) weakInstance = self;
  return ^(void (^transactionBlock)()) {
    StackTransaction *transaction = [weakInstance pushTransaction];
    
    transaction.transactionCompletionBlock = ^{
      [weakInstance popTransaction];
    };
    
    transaction.transactionBlock = transactionBlock;
    transaction.saveSynchronously();
  };  
}

- (void (^)(void (^)(), void (^)()))asyncTransaction
{
  __weak typeof(self) weakInstance = self;
  return ^(void (^transactionBlock)(), void (^completionBlock)()) {
    StackTransaction *transaction = [weakInstance pushTransaction];
    
    transaction.transactionCompletionBlock = ^{
      [weakInstance popTransaction];
      !completionBlock ?: completionBlock();
    };
    
    transaction.transactionBlock = transactionBlock;
    transaction.saveAsynchronously();
  };
}

@end

File: Pod/Classes/Stack.swift
/*
  Copyright © 2015 Shaps Mohsenin. All rights reserved.

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions are met:

  1. Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

  2. Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.

  THIS SOFTWARE IS PROVIDED BY SHAPS MOHSENIN `AS IS' AND ANY EXPRESS OR
  IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
  EVENT SHALL THE APP BUSINESS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
  INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
  OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
  ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import CoreData

private let StackThreadContextKey = "stack_context"

/// This class is responsible for listening for NSNotification's on context changes. It exists purely as a convenience, so that we don't have to subclass NSObject on our Stack!
class StackContextHandler: NSObject {
  
  unowned var stack: Stack
  
  init(stack: Stack) {
    self.stack = stack
  }
  
  // Called when an NSManagedObjectContext posts changes
  func contextDidSaveContext(note: NSNotification) {
    stack.contextDidSaveContext(note, contextHandler: self)
  }
  
}

/// A Stack is a CoreData wrapper that provides a type-safe implementation for reading and writing to CoreData
public final class Stack: CustomStringConvertible, Readable {
  
  /// The context handler that will listen for notifications. Only used when stackType = .ManualMerge
  private var contextHandler: StackContextHandler?
  
  /// Returns a string representation of the Stack's configuration
  public var description: String {
    return configuration.description
  }
  
  /// Holds onto the various Stacks in your application (1 or more)
  private static var stacks = { return [String: Stack]() }()
  
  /// The persistent coordinator associated with this stack
  private let coordinator: NSPersistentStoreCoordinator
  
  /// The configuration associated with this stack
  private let configuration: StackConfiguration
  
  /// The root context. Used only when stackType != .ManualMerge
  lazy var rootContext: NSManagedObjectContext = {
    let context = NSManagedObjectContext(concurrencyType: .PrivateQueueConcurrencyType)
    context.persistentStoreCoordinator = self.coordinator
    return context
  }()
  
  /// The main thread context.
  private lazy var mainContext: NSManagedObjectContext = {
    let context = NSManagedObjectContext(concurrencyType: .MainQueueConcurrencyType)
    if self.configuration.stackType == .ManualMerge {
      context.persistentStoreCoordinator = self.coordinator
    } else {
      context.parentContext = self.rootContext
    }
    
    return context
  }()
  
  /**
   Returns a context associated with the current thread. This function may return an existing context, however if none exist, it will create one, add it to the threadDictionary and return it
   
   - returns: An NSManagedObjectContext that can be used on the current thread
   */
  func currentThreadContext() -> NSManagedObjectContext {
    if NSThread.isMainThread() || configuration.stackType == .MainThreadOnly {
      return mainContext
    }
    
    if let context = NSThread.currentThread().threadDictionary[StackThreadContextKey] as? NSManagedObjectContext {
      return context
    }
    
    let context = NSManagedObjectContext(concurrencyType: .PrivateQueueConcurrencyType)
    
    switch configuration.stackType {
    case .ParentChild:
      context.parentContext = mainContext
    default:
      context.persistentStoreCoordinator = coordinator
    }
    
    NSThread.currentThread().threadDictionary[StackThreadContextKey] = context
    return context
  }
  
  /**
   Adds a new Stack
   
   - parameter name:      The name for this stack. Must be unique!
   - parameter configure: The configuration to apply to this Stack
   */
  public class func add(name: String, configure: ((_: StackConfiguration) -> ())?) {
    let config = StackConfiguration(name: name)
    configure?(config)
    
    let stack = Stack(config: config)
    stacks[name] = stack
  }
  
  /**
   Returns the entity name associated with the specified NSManagedObject class
   
   - parameter objectClass: The object
   
   - returns: The entity name for this managedObject class or nil if not found
   */
  func entityNameForManagedObjectClass(managedObjectClass: AnyClass) -> String? {
    let entities = coordinator.managedObjectModel.entitiesByName.values
    
    for entity in entities {
      if entity.managedObjectClassName == NSStringFromClass(managedObjectClass) {
        return entity.name
      }
    }
    
    return nil
  }
  
  public class func stack(named name: String) -> Stack? {
    guard let stack = stacks[name] else {
      if name != DefaultConfiguration.name {
        print("Stack: attempted to load a stack named '\(name)' that doesn't exist. Use Stack.add(name, config) first.")
      }
      return nil
    }
    
    return stack
  }
  
  private init(config: StackConfiguration) {
    self.configuration = config
    
    let model = NSManagedObjectModel.mergedModelFromBundles(nil)
    precondition(model != nil, "Stack: A valid NSManagedObjectModel must be provided!")
    coordinator = NSPersistentStoreCoordinator(managedObjectModel: model!)
    
    do {
      var storeType: String!
      
      switch config.persistenceType {
      case .Binary:
        storeType = NSBinaryStoreType
      case .MemoryOnly:
        storeType = NSInMemoryStoreType
      default:
        storeType = NSSQLiteStoreType
      }
      
      let filename = config.name
      let pathExtension = config.persistenceType == .SQLite ? ".sqlite" : ".bin"
      let path = NSURL(string: filename + pathExtension, relativeToURL: config.storeURL)
      try coordinator.addPersistentStoreWithType(storeType, configuration: nil, URL: path, options: config.storeOptions)
    } catch {
      print("Unable to add a persistent store for configuration: \(config)")
    }
    
    if config.stackType == .ManualMerge {
      contextHandler = StackContextHandler(stack: self)
      NSNotificationCenter.defaultCenter().addObserver(contextHandler!, selector: #selector(StackContextHandler.contextDidSaveContext(_:)), name: NSManagedObjectContextDidSaveNotification, object: nil)
    }
  }
  
  deinit {
    if configuration.stackType == .ManualMerge {
      NSNotificationCenter.defaultCenter().removeObserver(contextHandler!, name: NSManagedObjectContextDidSaveNotification, object: rootContext)
    }
  }
  
  func contextDidSaveContext(note: NSNotification, contextHandler: StackContextHandler) {
    if note.object === mainContext {
      return
    }
    
    if let info = note.userInfo {
      let userInfo = NSDictionary(dictionary: info)
      
      if let updated = userInfo.objectForKey(NSUpdatedObjectsKey) as? Set<NSManagedObject> {
        for object in updated {
          do { try mainContext.existingObjectWithID(object.objectID) } catch { }
        }
      }
    }

    dispatch_async(dispatch_get_main_queue()) { () -> Void in
      self.mainContext.mergeChangesFromContextDidSaveNotification(note)
    }
  }
  
  public func write(transaction: (transaction: Transaction) throws -> Void, completion: ((NSError?) -> Void)?) {
    let block: () -> () = { [unowned self] in
      do {
        try transaction(transaction: Transaction(stack: self, context: self.currentThreadContext()))
        self.currentThreadContext().save(true, completion: completion)
      } catch {
        completion?(error as NSError)
      }
    }
    
    currentThreadContext().performBlockAndWait(block)
  }
}

// MARK: OSX Additions

extension Stack {
  
#if os(OSX)
  
  /**
  Provided as a convenience for OSX only -- since a lot of bindings rely on having access to an NSManagedObjectContext
  
  - returns: The NSManagedObjectContext for use ONLY on the main queue
  */
  public func mainThreadContext() -> NSManagedObjectContext {
    return mainContext
  }
#endif
  
}

// MARK: Default Stack

extension Stack {
  
  /// Returns the default Stack name
  private static let DefaultName = "Default"
  
  /**
   Returns a default Stack instance that uses the default configuration. You can modify the configuration by calling Stack.configureDefaults() early in your applications lifecycle
   
   - returns: The default Stack
   */
  public static func defaultStack() -> Stack {
    if let stack = Stack.stack(named: DefaultName) {
      return stack
    }
    
    let stack = Stack(config: DefaultConfiguration)
    stacks[DefaultName] = stack
    return stack
  }
  
  /// Private: Returns the default Stack configuration
  private static var DefaultConfiguration: StackConfiguration = {
    return StackConfiguration(name: DefaultName)
  }()
  
  /**
   Provides a convenience function for configuring the default Stack before its initialized
   
   - parameter configure: The configuration to apply
   */
  public class func configureDefaults(configure: (_: StackConfiguration) -> ()) {
    let configuration = StackConfiguration(name: DefaultName)
    configure(configuration)
    DefaultConfiguration = configuration
  }
  
  
}




